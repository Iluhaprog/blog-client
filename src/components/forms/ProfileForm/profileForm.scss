@import '../../../vars.scss';

.profile-form {

    h2 {
        color: $text-color-3;
        margin: 0;
        font-weight: 400;
        margin-bottom: 20px;
    }

    .row {
        margin-bottom: 20px;
    }

    .drop-zone {
        min-width: 120px;
        margin-right: 20px;
    }

    .hint-box {
        width: 100%;
        margin: 10px;
    }

    .form__box {
        max-width: 100%;
        margin: 0;
        margin-right: 20px;   
        
        label {
            white-space: nowrap;
        }
    }

    textarea,
    .skills .form__box {
        width: calc(2/3*100% - 10px);
        margin-right: 30px;
    }

    textarea {
        display: block;
    }

    textarea + .hint-box, 
    .skills .form__box + .hint-box {
       width: calc(1/3*100% + 20px);
       margin-left: 0;
    }
}