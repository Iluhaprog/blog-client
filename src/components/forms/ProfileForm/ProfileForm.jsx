import React from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import { Row } from '../../containers';
import { SendButton } from '../../buttons';
import { DropZone, LabeledInput, Textarea } from '../../formElements';
import { HintBox } from '../../boxes';
import { Loader } from '../../loaders/Loader';
import './profileForm.scss';

function ProfileForm(props) {
    const { handleSubmit, isFetch } = props;
    return (
        <form onSubmit={handleSubmit} className='profile-form'>
            <h2>Account info</h2>
            <Row alignItems='c' justifyContent='sb'>
                <Field
                    name='avatar'
                    component={DropZone}
                />
                <HintBox>
                    <p>Drag image and drop in this box or click on this box</p>
                </HintBox>
            </Row>
            <Row alignItems='c' justifyContent='sb'>
                <LabeledInput 
                    type='text'
                    name='firstName'
                    component='input'
                    label='First Name'
                />
                <LabeledInput 
                    type='text'
                    name='lastName'
                    component='input'
                    label='Last Name'
                />
                <HintBox>
                    <p>Write your real first name and last name</p>
                </HintBox>
            </Row>
            <Row alignItems='c' justifyContent='sb'>
                <LabeledInput 
                    type='text'
                    name='username'
                    component='input'
                    label='Username'
                />
                <LabeledInput 
                    type='text'
                    name='email'
                    component='input'
                    label='Email'
                />
                <HintBox>
                    <p>Write your real email namd and unique username</p>
                </HintBox>
            </Row>
            <h2>Profile info</h2>
            <Row alignItems='c' justifyContent='sb'>
                <Textarea name='bio' placeholder='Bio' />
                <HintBox>
                    <p>Write about your developer path, projects etc. </p>
                </HintBox>
            </Row>
            <div className='skills'>
                <Row alignItems='c' justifyContent='sb'>
                    <LabeledInput 
                        type='text'
                        name='skills'
                        component='input'
                        label='Skills'
                        />
                    <HintBox>
                        <p>Write the skills that you own, separated by commas</p>
                    </HintBox>
                </Row>
            </div>
            <Row justifyContent='fs'>
                <SendButton text='Apply' />
                <Loader
                    visible={isFetch}
                />
            </Row>
        </form>
    );
}

const mapStateToProps = state => ({
    initialValues: {
        avatar: state.user.avatarImage,
        firstName: state.user.firstName,
        lastName: state.user.lastName,
        username: state.user.username,
        email: state.user.email,
        bio: state.user.bio,
        skills: state.user.skills,
    },
    isFetch: state.user.isFetch,
});

export default connect(mapStateToProps)(reduxForm({
    form: 'profile',
    enableReinitialize: true,
})(ProfileForm));