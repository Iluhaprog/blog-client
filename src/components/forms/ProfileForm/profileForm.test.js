import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux';

import ProfileForm from './ProfileForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const defaultUserState = {
    user : { 
        avatarImage: 'img.jpg',
        firstName: 'Test fn',
        lastName: 'Test ln',
        username: 'test un',
        bio: 'Test bio',
        email: 'test@test.test',
        skills: 'test skills',
        isFetch: false,
    },
};

describe('Test for profile form view', () => {
    test('Should have values from inital state', async () => {
        const store = mockStore(defaultUserState);
        const { asFragment } = render(
            <Provider store={store}>
                <ProfileForm onSubmit={e => e} />
            </Provider>
        );

        await waitFor(() => {
            expect(!!store.getActions().length).toBe(true);
            expect(asFragment()).toMatchSnapshot();
        });
    });
});

describe('Test for profile form submit', () => {
    test('Should make submit', async () => {
        const store = mockStore(defaultUserState);
        const submit = jest.fn();
        const { asFragment, container } = render(
            <Provider store={store}>
                <ProfileForm onSubmit={submit} />
            </Provider>
        );
        
        fireEvent.submit(container.querySelector('form'));

        await waitFor(() => {
            expect(submit).toHaveBeenCalledTimes(1);
        });
    });
});