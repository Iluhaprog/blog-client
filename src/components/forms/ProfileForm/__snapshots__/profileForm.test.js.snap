// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for profile form view Should have values from inital state 1`] = `
<DocumentFragment>
  <form
    class="profile-form"
  >
    <h2>
      Account info
    </h2>
    <div
      class="row row_jc-sb row_ai-c "
    >
      <div
        class="drop-zone "
      >
        <img
          alt="empty"
          class="empty"
          src="picture.svg"
        />
        <input
          type="file"
        />
      </div>
      <div
        class="hint-box"
      >
        <p>
          Drag image and drop in this box or click on this box
        </p>
      </div>
    </div>
    <div
      class="row row_jc-sb row_ai-c "
    >
      <div
        class="form__box row-reverse"
      >
        <input
          name="firstName"
          type="text"
          value=""
        />
        <label
          for="firstName"
        >
          First Name
        </label>
      </div>
      <div
        class="form__box row-reverse"
      >
        <input
          name="lastName"
          type="text"
          value=""
        />
        <label
          for="lastName"
        >
          Last Name
        </label>
      </div>
      <div
        class="hint-box"
      >
        <p>
          Write your real first name and last name
        </p>
      </div>
    </div>
    <div
      class="row row_jc-sb row_ai-c "
    >
      <div
        class="form__box row-reverse"
      >
        <input
          name="username"
          type="text"
          value=""
        />
        <label
          for="username"
        >
          Username
        </label>
      </div>
      <div
        class="form__box row-reverse"
      >
        <input
          name="email"
          type="text"
          value=""
        />
        <label
          for="email"
        >
          Email
        </label>
      </div>
      <div
        class="hint-box"
      >
        <p>
          Write your real email namd and unique username
        </p>
      </div>
    </div>
    <h2>
      Profile info
    </h2>
    <div
      class="row row_jc-sb row_ai-c "
    >
      <textarea
        class="textarea"
        name="bio"
        placeholder="Bio"
      />
      <div
        class="hint-box"
      >
        <p>
          Write about your developer path, projects etc. 
        </p>
      </div>
    </div>
    <div
      class="skills"
    >
      <div
        class="row row_jc-sb row_ai-c "
      >
        <div
          class="form__box row-reverse"
        >
          <input
            name="skills"
            type="text"
            value=""
          />
          <label
            for="skills"
          >
            Skills
          </label>
        </div>
        <div
          class="hint-box"
        >
          <p>
            Write the skills that you own, separated by commas
          </p>
        </div>
      </div>
    </div>
    <div
      class="row row_jc-fs  "
    >
      <button
        class="button send-button"
      >
        Apply
      </button>
      <img
        alt="loading..."
        class="loader "
        height="30"
        src="loader.svg"
        width="30"
      />
    </div>
  </form>
</DocumentFragment>
`;
